File: js/2017/04-HighEntropyPassphrases/solution.js
const fs = require("fs");
const input = fs.readFileSync("input.txt", "ascii");
const passphrases = input.split(/\r?\n/);

function solution_part1() {
	let valid_passphrases = passphrases.length; // Innocent until proven guilty

	for (passphrase of passphrases) {
		let words = passphrase.split(" ");
		for (let i = 0; i < words.length; i++) {
			// Check if first occurance of word is at a different index
			if (words.indexOf(words[i]) !== i) {
				valid_passphrases--;
				break;
			}
		}
	}

	return valid_passphrases;
}

function is_anagram(first, second) {
	if (first.length !== second.length) return false;
	// JS's string.prototype.replace() replaces only the first instance, making it suitable for this.
	for (char of first) second = second.replace(char, "");
	return second.length === 0;
}

function solution_part2() {
	let valid_passphrases = passphrases.length;

	outer: for (passphrase of passphrases) {
		let words = passphrase.split(" ");
		for (let i = 0; i < words.length; i++) {
			for (let j = 0; j < words.length; j++) {
				if (i === j) continue;
				if (is_anagram(words[i], words[j])) {
					valid_passphrases--;
					continue outer;
				}
			}
		}
	}

	return valid_passphrases;
}

console.log("Solution Part 1: " + solution_part1());
console.log("Solution Part 2: " + solution_part2());

File: py/2017/04-HighEntropyPassphrases/solution.py
import os

with open("input.txt", "r") as file:
	input = file.read()

passphrases = input.splitlines()

def solution_part1():
	valid_passphrases = 0

	for passphrase in passphrases:
		valid_passphrases += 1 # Innocent until proven guilty
		words = passphrase.split(" ")
		word_instances = []
		for word in words:
			if word in word_instances:
				valid_passphrases -= 1
				break
			word_instances.append(word)
	return valid_passphrases

def solution_part2():
	valid_passphrases = 0

	for passphrase in passphrases:
		valid = True
		words = passphrase.split(" ")
		words_to_compare = []
		for word in words:
			for compared_word in words_to_compare:
				if len(word) != len(compared_word):
					continue
				word_chars = word
				for char in compared_word:
					word_chars = word_chars.replace(char, "", 1)
				if len(word_chars) == 0:
					valid = False
					break
			if not valid:
				break
			words_to_compare.append(word)
		if valid:
			valid_passphrases += 1
	return valid_passphrases


print("Solution Part 1:", solution_part1())
print("Solution Part 2:", solution_part2())